runtime/src/main/cpp/Exceptions.cpp

```diff
@@ -227,79 +226,110 @@ void OnUnhandledException(KRef throwable) {
+      }
+      _Exit(EXIT_FAILURE); // force exit
+    }
+} concurrentTerminateWrapper;
 
+void reportUnhandledException(KRef throwable) {
+  OnUnhandledException(throwable);
 #if KONAN_REPORT_BACKTRACE_TO_IOS_CRASH_LOG
-  {
-    LockGuard<SimpleMutex> lock(terminatingMutex);
-    if (!terminating) {
-      ReportBacktraceToIosCrashLog(throwable);
-    }
-  }
+  ReportBacktraceToIosCrashLog(throwable);
 #endif
+}
+
+} // namespace
 
-  konan::abort();
+RUNTIME_NORETURN void TerminateWithUnhandledException(KRef throwable) {
+  concurrentTerminateWrapper([=]() {
+    reportUnhandledException(throwable);
+    konan::abort();
+  });
 }
 
 // Some libstdc++-based targets has limited support for std::current_exception and other C++11 functions.
 // This restriction can be lifted later when toolchains will be updated.
 #if KONAN_HAS_CXX11_EXCEPTION_FUNCTIONS
 
-static void (*oldTerminateHandler)() = nullptr;
-
-static void callOldTerminateHandler() {
-#if KONAN_REPORT_BACKTRACE_TO_IOS_CRASH_LOG
-  {
-    LockGuard<SimpleMutex> lock(terminatingMutex);
-    terminating = true;
+namespace {
+class TerminateHandler {
+
+	// In fact, it's safe to call my_handler directly from outside: it will do the job and then invoke original handler,
+	// even if it has not been initialized yet. So one may want to make it public and/or not the class member
+  RUNTIME_NORETURN static void kotlinHandler() {
+    concurrentTerminateWrapper([]() {
+      if (auto currentException = std::current_exception()) {
+        try {
+          std::rethrow_exception(currentException);
+        } catch (ExceptionObjHolder& e) {
+          reportUnhandledException(e.obj());
+          konan::abort();
+        } catch (...) {
+          // Not a Kotlin exception - call default handler
+          instance()->queuedHandler_();
+        }
+      }
+      // Come here in case of direct terminate() call or unknown exception - go to default terminate handler.
+      instance()->queuedHandler_();
+    });
   }
-#endif
 
-  RuntimeCheck(oldTerminateHandler != nullptr, "Underlying exception handler is not set.");
-  oldTerminateHandler();
-}
+  using QH = __attribute__((noreturn)) void(*)();
+  QH queuedHandler_;
 
-static void KonanTerminateHandler() {
-  auto currentException = std::current_exception();
-  if (!currentException) {
-    // No current exception.
-    callOldTerminateHandler();
-  } else {
-    try {
-      std::rethrow_exception(currentException);
-    } catch (ExceptionObjHolder& e) {
-      TerminateWithUnhandledException(e.obj());
-    } catch (...) {
-      // Not a Kotlin exception.
-      callOldTerminateHandler();
-    }
+  /// Use machinery like Meyers singleton to provide thread safety
+  TerminateHandler()
+    : queuedHandler_((QH)std::set_terminate(kotlinHandler)) {}
+
+  static TerminateHandler* instance() {
+    static TerminateHandler* singleton [[clang::no_destroy]] = new TerminateHandler();
```

Comment:
Are pointers necessary here still? Because I don't think `no_destroy` means anything for primitive types.